### DISCUSSION TITLE
How can Tell which backend am I using?

### DISCUSSION THREAD

**Discussion by plum** [tramhao/termusic#373] [Q&A]
I see there are alternative backends for termusic. And from the source I see in build configs,  for example , 'backend=rusty'.

I expect in the archlinux/aur build files, or the git file, more than one backend is compiled.
 
Q.1. Given the above,  how do I find which one I am running,, or how is one selected over another in a config file?

Q.2. Secondarily and related,  without myself specifying a preferred backend, which backend is preferred by termusic assuming more than one is available?

Thanks,
John

**Comment by tramhao** [comment]
We could compile all backends and use the command line flag to choose which backend to use. The default is rusty backend. Currently there is no config option for this yet.

**Comment by plum** [comment]
@tramhao  thanks for kind reply. I see that its something to be comiled. Is thre any advantage in using mpv, for example, more  control over playback?

**Comment by tramhao** [comment]
For mpv, the only advantage I think is that when listening to podcasts, speed up is more natural than gstreamer and rusty backend. It'll keep the tone. For gstreamer and rusty, the tone is kept also, but not as good as mpv.

**Comment by hasezoey** [comment]
> Q.1. Given the above, how do I find which one I am running,, or how is one selected over another in a config file?

unless you look at the `--backend` on the cli, there are only 2 ways to determine which backend you are using:
look in the log for `INFO  termusicplayback]: Using Backend "THE BACKEND"` or look in your sounds config / Mixer what name termusic shows up as (example in kde+pipewire and rusty backend `Pipewire ALSA [termusic-server]: ALSA playback`)

> Q.2. Secondarily and related, without myself specifying a preferred backend, which backend is preferred by termusic assuming more than one is available?

at least on current master, [the `rusty` backend is always compiled in and is the default unless another is selected](https://github.com/tramhao/termusic/blob/8aed705850ddc6022be3baaed5e3dcabf2e8ba9f/server/src/cli.rs#L47-L48).
Note that there currently is no config option to specify the backend in the config.
(previously before 211fc3fe008932d052d31d3b836e8a80eade3cfe, default was `rusty`, but could also be specified to not be compiled)

> I expect in the archlinux/aur build files, or the git file, more than one backend is compiled.

[extra/termusic](https://gitlab.archlinux.org/archlinux/packaging/packages/termusic/-/blob/main/PKGBUILD?ref_type=heads) compiles the default, which currently is only `rusty`
[aur/termusic-git](https://aur.archlinux.org/packages/termusic-git) compiles all backends (via `--all-features`)

---

PS: as for advantages over other backends that i can think of:

rusty:
- purely rust, you only need ALSA (unless you also have `rusty-soundtouch`, then you need the `soundtouch` library)
- comparatively low codec support
- buffering is handled by us, and compared to the other backends is likely not the best

mpv:
- relies on libmpv, which relies on ffmpeg, and so has a wide variety of codec support

gst:
- relies on gstreamer, which also has a wide variety of codec support

also note that between those backends, the volume handling might be different (for example on my machine gst and mpv are way louder than rusty)

**Reply by hasezoey** [reply]
[Since 0.12.0](https://github.com/tramhao/termusic/blob/master/CHANGELOG.md#v0120) the backend can now be configured in the config editor. Note though that it will show *all backends* not just the available ones in the server, potentially causing the next startup to fail with "unavailable backend".
